app/unit-tests/src/org/commcare/android/nfc/NfcManagerTest.java

```diff
@@ -0,0 +1,74 @@
+
+        // Empty payload should not have any tag attached
+        assert nfcManager.tagAndEncryptPayload("").contentEquals("");
+
+        // Check for right tag
+        String encryptedMessage = nfcManager.tagAndEncryptPayload(PAYLOAD);
+        assert encryptedMessage.startsWith(NFC_ENCRYPTION_SCHEME + PAYLOAD_DELIMITER + ENTITY_ID + PAYLOAD_DELIMITER);
+
+        // decrypt message and ensure we get the same payload
+        assert nfcManager.decryptValue(encryptedMessage).contentEquals(PAYLOAD);
+    }
+
+    @Test
+    public void emptyEncryptionKeyTest() throws EncryptionUtils.EncryptionException {
+        NfcManager nfcManager = new NfcManager(CommCareTestApplication.instance(), "", ENTITY_ID);
+        String encryptedMessage = nfcManager.tagAndEncryptPayload(PAYLOAD);
+        assert encryptedMessage.startsWith(PAYLOAD_DELIMITER + ENTITY_ID + PAYLOAD_DELIMITER);
+        assert nfcManager.decryptValue(encryptedMessage).contentEquals(PAYLOAD);
+    }
+
+    @Test
+    public void emptyEntityIdTest() throws EncryptionUtils.EncryptionException {
+        NfcManager nfcManager = new NfcManager(CommCareTestApplication.instance(), ENCRYPTION_KEY, "");
+        String encryptedMessage = nfcManager.tagAndEncryptPayload(PAYLOAD);
+        assert encryptedMessage.startsWith(NFC_ENCRYPTION_SCHEME + PAYLOAD_DELIMITER + PAYLOAD_DELIMITER);
+        assert nfcManager.decryptValue(encryptedMessage).contentEquals(PAYLOAD);
+    }
+
+    @Test
+    public void emptyEncryptionKeyAndEntityIdTest() throws EncryptionUtils.EncryptionException {
+        NfcManager nfcManager = new NfcManager(CommCareTestApplication.instance(), "", "");
+        String encryptedMessage = nfcManager.tagAndEncryptPayload(PAYLOAD);
+        assert encryptedMessage.startsWith(PAYLOAD_DELIMITER + PAYLOAD_DELIMITER);
+        assert nfcManager.decryptValue(encryptedMessage).contentEquals(PAYLOAD);
+    }
+
+    @Test(expected = NfcManager.InvalidPayloadTagException.class)
+    public void readingPayloadWithDifferentTag_shouldFail() throws EncryptionUtils.EncryptionException {
+        NfcManager nfcManager = new NfcManager(CommCareTestApplication.instance(), ENCRYPTION_KEY, "some_other_id");
+        String encryptedMessage = nfcManager.tagAndEncryptPayload(PAYLOAD);
+        new NfcManager(CommCareTestApplication.instance(), ENCRYPTION_KEY, ENTITY_ID)
+                .decryptValue(encryptedMessage).contentEquals(PAYLOAD);
+    }
+
+    @Test
+    public void payloadWithoutTagTest() throws EncryptionUtils.EncryptionException {
+        // Decrypt an old payload without specifying encryptionKey and entityId
+        assert new NfcManager(CommCareTestApplication.instance(), "", "")
+                .decryptValue(PAYLOAD).contentEquals(PAYLOAD);
+
+        // Decrypt an old payload with some encryptionKey and entityId
+        assert new NfcManager(CommCareTestApplication.instance(), ENCRYPTION_KEY, ENTITY_ID)
```

Comment:
@ctsims This test currently fails with a `InvalidPayloadTagException`. I am unsure what should be right behaviour here when you try to read a payload written on old CC versions with a NFC callout that has specified an `encryptionKey` and `entityID`. Is it a right expectation to set that if you specify one of `encryptionKey` and `entityID`, you expect to only read payloads that were written with these exact params in the write callout.

Reply:
As per our conversation, I think it's fine to either fail in the case you are describing or to gracefully update a written entity which doesn't have a version.

I definitely think if someone has a record written with an `entityID` though, that if there's another record with a different entity ID we should have clearly documented expectations about the failure mode.

Reply:
Does [this](https://github.com/dimagi/commcare-android/pull/2502/commits/a38384d516e428e688026610c62c7730b2175424) like an ok upgrade pathway ?

> I definitely think if someone has a record written with an entityID though, that if there's another record with a different entity ID we should have clearly documented expectations about the failure mode.

Can you elaborate a little more on this. Today this case will fail as long as the `ignore_tag` parameter introduced in last commit is `false` but user have an option to read the whole payload as it is when `ignore_tag` is set to `true`.